glean-core/src/ping/mod.rs

```diff
@@ -222,11 +222,45 @@ impl PingMaker {
     ) -> Option<Ping<'a>> {
         info!("Collecting {}", ping.name());
 
-        let metrics_data = StorageManager.snapshot_as_json(glean.storage(), ping.name(), true);
+        let mut metrics_data = StorageManager.snapshot_as_json(glean.storage(), ping.name(), true);
         let events_data = glean
             .event_storage()
             .snapshot_as_json(glean, ping.name(), true);
 
+        // Due to the way the experimentation identifier could link datasets that are intentionally unlinked,
+        // it will not be included in pings that specifically exclude the Glean client-id.
+        if (!ping.include_client_id() || !ping.send_if_empty())
+            && glean.test_get_experimentation_id().is_some()
+            && metrics_data.is_some()
+        {
+            // There is a lot of unwrapping here, but that's fine because the `if` conditions above mean that the
+            // experimentation id is present in the metrics.
+            let metrics = metrics_data.as_mut().unwrap().as_object_mut().unwrap();
+            let metrics_count = metrics.len();
+            let strings = metrics.get_mut("string").unwrap().as_object_mut().unwrap();
+            let string_count = strings.len();
+
+            if !ping.include_client_id() {
+                strings.remove("glean.client.annotation.experimentation_id");
+            } else {
+                // Handle the send_if_empty case by checking if this is the only metric in the data. If this is the only
+                // metric in the data, and this is a ping that should not be sent if empty, then we should remove the
+                // experimentation id metric so `is_empty` below is correct.
+                if metrics_count == 1 && string_count == 1 {
+                    //
+                    strings.remove("glean.client.annotation.experimentation_id");
+                }
```

Comment:
What about this logic instead:

```suggestion
            let empty_payload = events_data.is_none() && metrics_count == 1 && string_count == 1;

            if !ping.include_client_id() || (!ping.send_if_empty() && empty_payload) {
                strings.remove("glean.client.annotation.experimentation_id");
            }
```